pkg/controller/windowsmachineconfig/tracker.go

```diff
@@ -78,14 +75,110 @@ func newNodeRecord(ipAddress string, secretName string) nodeRecord {
 	}
 }
 
+// getWindowsVM returns a windowsVM interface to be used for running commands against
+func getWindowsVM(ipAddress, instanceID string, credentials Credentials) (types.WindowsVM, error) {
+	winVM := &types.Windows{}
+	windowsCredentials := types.NewCredentials(instanceID, ipAddress, credentials.Password, credentials.Username)
+	winVM.Credentials = windowsCredentials
+	// Set up Winrm client
+	if err := winVM.SetupWinRMClient(); err != nil {
+		return nil, errors.Wrap(err, "error instantiating winrm client")
+	}
+	// Set up SSH client
+	if err := winVM.GetSSHClient(); err != nil {
+		return nil, errors.Wrap(err, "error instantiating ssh client")
+	}
+	return winVM, nil
+}
+
+// getInstanceIP gets the instance IP address associated with a node
+func getInstanceIP(nodeList *v1.NodeList, instanceID string) (string, error) {
+	// Ignore the nodes that are not ready.
+	for _, node := range nodeList.Items {
+		for _, condition := range node.Status.Conditions {
+			if condition.Type != v1.NodeReady {
+				continue
+			}
+		}
+		if strings.Contains(node.Spec.ProviderID, instanceID) {
+			for _, address := range node.Status.Addresses {
+				// If external ip exists return it as it is used in AWS, if it doesn't return internal(in case of azure)
+				// TODO: collect all the ips and define a priority order, external then internal etc.
+				if address.Type == v1.NodeExternalIP && len(address.Address) > 0 {
+					return address.Address, nil
+				}
+			}
+		}
+	}
+	return "", errors.New("unable to find Windows Worker nodes")
+}
+
+// getCredsFromSecret gets the credentials associated with the instance.
+func getCredsFromSecret(k8sclientset *kubernetes.Clientset, instanceID string, operatorNS string) (Credentials, error) {
+	creds := Credentials{}
+	instanceSecret, err := k8sclientset.CoreV1().Secrets(operatorNS).Get(instanceID, metav1.GetOptions{})
+	if err != nil && k8serrors.IsNotFound(err) {
+		return creds, errors.Wrap(err, "getting secret")
+	}
+	encodedCreds := instanceSecret.Data[instanceID]
+	if err := json.Unmarshal(encodedCreds, &creds); err != nil {
+		return creds, errors.Wrap(err, "unmarshalling creds failed")
+	}
+	return creds, err
+}
+
+// initWindowsVMs initializes the windowsVMs map from the store configmMap and corresponding secrets
+func initWindowsVMs(k8sclientset *kubernetes.Clientset, operatorNS string) (map[types.WindowsVM]bool, error) {
+	windowsVMs := make(map[types.WindowsVM]bool)
```

Comment:
Please create the map close to the scope of it usage. If you create it after you get the list of nodes, you can specify a size.

Reply:
The link does not work but I think understand what you are implying. I would just pass an empty map below. But I am fine either way.